### PR TITLE
Adds tosa reduce_sum to rock reduce conversion

### PR DESCRIPTION
**Pull request by manupak**
This commit adds support for converting tosa.reduce_sum to rock reduce.

Moreover, a new rock attribute (maybe temporary) to indicate args of kernels where it requires to be initialized to some value prior to launch. It is verified by doing the same in the clone verification flow by adding linalg.fills.

closes : https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/791

### CODE REVIEW

**Comment by manupak** [comment]
I'd appreciate a review here to move this forward.

**Comment by manupak** [comment]
@sjw36 would you be able to take a look at this ? I need this to be able to use tosa + clone flow for verification.

File: mlir/lib/Dialect/Rock/Transforms/AffixTuningParameters.cpp [hunk 1]
```diff
@@ -58,6 +58,13 @@ void AffixTuningParameters::runOnOperation() {
 
   func.walk(
       [&](RockGemmWrapperInterface op) { affixTuningParametersImpl(op); });
+  func.walk([&](ReduceOp op) {
+    func::FuncOp funcOp = getOperation();
```
Comment (sjw36) [inline mlir/lib/Dialect/Rock/Transforms/AffixTuningParameters.cpp:62]: This `funcOp` is the same as `func` in the captured scope.

File: mlir/lib/Dialect/Rock/Transforms/AffixTuningParameters.cpp [hunk 2]
```diff
@@ -58,6 +58,13 @@ void AffixTuningParameters::runOnOperation() {
...
+    if (!funcOp->hasAttr("block_size")) {
+      funcOp->setAttr("block_size", op.getBlockSizeAttr());
+      funcOp->setAttr("grid_size", op.getGridSizeAttr());
```
Comment (sjw36) [inline mlir/lib/Dialect/Rock/Transforms/AffixTuningParameters.cpp:65]: what happens if these are already set on the func, how do you resolve?

File: mlir/test/Conversion/TosaToRock/reductions/tosa-to-rock-reduce-e2e.mlir [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// RUN: rocmlir-opt --tosa-partition="anchor-ops=tosa.reduce_sum" --xmodel-async-graph --xmodel-target-kernels="targets=%arch" %s | rocmlir-driver -host-pipeline highlevel | rocmlir-gen -ph -print-results -rand 1 -rand_type float -fut test_basic --verifier clone - | rocmlir-driver -host-pipeline xmodel -kernel-pipeline full -targets %arch | xmir-runner --shared-libs=%linalg_test_lib_dir/libmlir_rocm_runtime%shlibext,%conv_validation_wrapper_library_dir/libconv-validation-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext,%linalg_test_lib_dir/libmlir_c_runner_utils%shlibext --entry-point-result=void | FileCheck %s --check-prefix=CLONE
+// cat %s | rocmlir-driver -host-pipeline xmodel -kernel-pipeline full -targets %arch | xmir-runner --shared-libs=%linalg_test_lib_dir/libmlir_rocm_runtime%shlibext,%conv_validation_wrapper_library_dir/libconv-validation-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext,%linalg_test_lib_dir/libmlir_c_runner_utils%shlibext --entry-point-result=void | FileCheck %s --check-prefix=CLONE
```
Comment (sjw36) [inline mlir/test/Conversion/TosaToRock/reductions/tosa-to-rock-reduce-e2e.mlir:2]: what is this line for?

**Review by sjw36 — APPROVED** [review]
Looks good. Just some minor comments/questions.

File: mlir/lib/Dialect/Rock/Transforms/AffixTuningParameters.cpp (hunk 1 above)
Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/AffixTuningParameters.cpp:62]: ah yeah! silly me. removed it.

File: mlir/lib/Dialect/Rock/Transforms/AffixTuningParameters.cpp (hunk 2 above)
Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/AffixTuningParameters.cpp:65]: its based on precedence and right now gemm like ops takes precedence (if they are present and that scenario represent a fusion case).  What I've codified here is the walk order by inserting the reduction op walker after gemm-like op walker.

File: mlir/test/Conversion/TosaToRock/reductions/tosa-to-rock-reduce-e2e.mlir (hunk 3 above)
Comment (manupak) [reply mlir/test/Conversion/TosaToRock/reductions/tosa-to-rock-reduce-e2e.mlir:2]: shouldn't have been there.. removed it. Thanks for spotting.

**Comment by manupak** [comment]
Thanks @sjw36!

**Pull request merged**